File: 백준/Silver/2468. 안전 영역/안전 영역.py
import sys

input = sys.stdin.readline

N = int(input())
graph = []
for _ in range(N):
    _list = list(map(int, input().split()))
    graph.append(_list)

max_precipitations = []
min_precipitations = []
for _graph in graph:
    max_precipitations.append(max(_graph))
    min_precipitations.append(min(_graph))


def bfs(a, b, c):
    queue = []
    queue.append((a, b))

    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    while queue:
        x, y = queue.pop(0)

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < N and 0 <= ny < N:
                if graph[nx][ny] > c and visited[nx][ny] == 0:
                    visited[nx][ny] = 1
                    queue.append((nx, ny))


count = 1
for i in range(min(min_precipitations), max(max_precipitations) + 1):
    _count = 0
    visited = [[0 for _ in range(N)] for _ in range(N)]
    for x in range(N):
        for y in range(N):
            if graph[x][y] > i and visited[x][y] == 0:
                visited[x][y] = 1
                bfs(x, y, i)
                _count += 1
    count = max(count, _count)

print(count)

File: 백준/Silver/2468. 안전 영역/안전 영역.js
const readline = require("readline");
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let inputLines = [];
rl.on("line", (line) => {
  inputLines.push(line);
});

rl.on("close", () => {
  const N = parseInt(inputLines[0]);
  const safeZone = [];
  for (let i = 1; i <= N; i++) {
    safeZone.push(inputLines[i].split(" ").map(Number));
  }

  const visited = new Array(N).fill(0).map(() => new Array(N).fill(0));

  const dx = [-1, 1, 0, 0];
  const dy = [0, 0, -1, 1];

  function dfs(x, y, rainfall) {
    for (let i = 0; i < 4; i++) {
      const nx = x + dx[i];
      const ny = y + dy[i];
      if (
        nx >= 0 &&
        nx < N &&
        ny >= 0 &&
        ny < N &&
        safeZone[nx][ny] > rainfall &&
        visited[nx][ny] === 0
      ) {
        visited[nx][ny] = 1;
        dfs(nx, ny, rainfall);
      }
    }
  }

  const rainfallArray = [];
  for (let i = 0; i < N; i++) {
    for (let j = 0; j < N; j++) {
      rainfallArray.push(safeZone[i][j]);
    }
  }

  const minRainfall = Math.min(...rainfallArray);
  const maxRainfall = Math.max(...rainfallArray);

  let count = 1;

  for (let rainfall = minRainfall; rainfall < maxRainfall; rainfall++) {
    let _count = 0;
    visited.forEach((row) => row.fill(0));
    for (let i = 0; i < N; i++) {
      for (let j = 0; j < N; j++) {
        if (safeZone[i][j] > rainfall && visited[i][j] === 0) {
          visited[i][j] = 1;
          dfs(i, j, rainfall);
          _count += 1;
        }
      }
    }
    count = Math.max(count, _count);
  }

  console.log(count);
});
